compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1808,9 +1808,18 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         candidate.visit_leaves(|leaf_candidate| {
             last_otherwise = leaf_candidate.otherwise_block;
         });
+
         let remaining_match_pairs = mem::take(&mut candidate.match_pairs);
+        debug_assert!(
+            remaining_match_pairs
+                .iter()
+                .all(|match_pair| matches!(match_pair.test_case, TestCase::Or { .. }))
+        );
+
         candidate.visit_leaves(|leaf_candidate| {
             assert!(leaf_candidate.match_pairs.is_empty());
+            // The remaining match pairs are all `Or`, so we can append them
+            // without having to re-sort or-patterns to the end.
```

Comment:
We're adding them to an empty vector, so there would be no need to sort them anyway. I made sure that no part of the algorithm relies on or-patterns being sorted to the end: it is purely a performance optimization.

Reply:
I guess it would have been clearer to write
```rust
leaf_candidate.match_pairs = remaining_match_pairs.iter().cloned().collect();
```

Reply:
Oh you're right; there's even a relevant assertion on the line above that I somehow missed.

I think the main thing here is to rewrite my comment to be more aware of what's actually happening, and to add better spacing/grouping.